test/user_test.dart

```diff
@@ -0,0 +1,90 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////////
+
+import 'dart:io';
+
+import 'package:test/expect.dart';
+
+import '../lib/realm.dart';
+import 'test.dart';
+
+Future<void> main([List<String>? args]) async {
+  print("Current PID $pid");
+
+  await setupTests(args);
+
+  baasTest('User logout anon user is marked as removed', (configuration) async {
+    final app = App(configuration);
+    final user = await app.logIn(Credentials.anonymous());
+    expect(user.state, UserState.loggedIn);
+    await user.logout();
+    expect(user.state, UserState.removed);
+  });
+
+  baasTest('User logout', (configuration) async {
+    final app = App(configuration);
+    final user = await app.logIn(Credentials.emailPassword(testUsername, testPassword));
+    expect(user.state, UserState.loggedIn);
+    await user.logout();
+    expect(user.state, UserState.loggedOut);
+  });
+
+  baasTest('User get id', (configuration) async {
+    final app = App(configuration);
+    final user = await app.logIn(Credentials.emailPassword(testUsername, testPassword));
+    expect(user.id, isNotEmpty);
+  });
+
+  baasTest('User get identities', (configuration) async {
+    final app = App(configuration);
+    final user = await app.logIn(Credentials.emailPassword(testUsername, testPassword));
+    expect(user.identities.any((identity) => identity.provider == AuthProviderType.emailPassword), isTrue);
+  }, skip: "Blocked on https://github.com/realm/realm-core/issues/5467");
+
+  baasTest('User get customdata', (configuration) async {
+    final app = App(configuration);
+    final user = await app.logIn(Credentials.emailPassword(testUsername, testPassword));
+    expect(user.customData, isNull);
+  });
+
+  baasTest('User refresh customdata', (configuration) async {
+    final app = App(configuration);
+    final user = await app.logIn(Credentials.emailPassword(testUsername, testPassword));
+    final dynamic data = await user.refreshCustomData();
+    expect(data, isNull);
+  });
+
+  baasTest('User link credentials', (configuration) async {
+    final app = App(configuration);
+    final user = await app.logIn(Credentials.anonymous());
+  
+    expect(user.state, UserState.loggedIn);
+    expect(user.identities.length, 1);
+    expect(user.identities.singleWhere((identity) => identity.provider == AuthProviderType.anonymous).provider, isNotNull);
+
+    final authProvider = EmailPasswordAuthProvider(app);
```

Comment:
Must have missed that when reviewing the PR that adds email password auth, but why do we have a constructor like that? I think we should stick to the API exposed by other SDKs and expose the auth provider as a property on the app - like `app.emailPasswordAuth`.

Reply:
Why do we have a special case for this provider versus the others. Why not allow constructing it outside the app.

Reply:
It's mostly for discoverability. All the static "clients" - i.e. emailpassword, sync - are exposed as properties on App and all the user-scoped clients - i.e. push, remotemongo, functions - are exposed on User.